### ISSUE TITLE
"IPPool.spec.nodeSelector: '!all()'" is invalid

### ISSUE THREAD

**Issue by yell93** [projectcalico/calico#8823]
## Expected Behavior

After setting `IPPool.spec.nodeSelector: '!all()'`, must I manually specify `cni.projectcalico.org/ipv4pools: '[" poolName "]`, otherwise the calico-ipam cannot use this `poolName` to automatically assign `IP`

As stated in the documentation: 
![image](https://github.com/projectcalico/calico/assets/16219420/fb377a62-0bf1-4c8d-bb04-7d7ba7892efe)

## Current Behavior

```
- apiVersion: crd.projectcalico.org/v1
  kind: IPPool
  metadata:
    name: pool-10
  spec:
    blockSize: 26
    cidr: 10.192.230.128/26
    ipipMode: Never
    natOutgoing: false
    nodeSelector: '!all()'
```
pool-10 is set, but calico-ipam can still automatically allocate pool-10's cidr(`10.192.230.128/26`), and automatically allocate IP.

```
[root@master-23to24-199-146-28 ~]# calicoctl ipam show --show-blocks
+----------+-------------------+-----------+------------+------------+
| GROUPING |       CIDR        | IPS TOTAL | IPS IN USE |  IPS FREE  |
+----------+-------------------+-----------+------------+------------+
| IP Pool  | 10.192.228.0/22   |      1024 | 37 (4%)    | 987 (96%)  |
| Block    | 10.192.230.128/26 |        64 | 3 (5%)     | 61 (95%)   |
| Block    | 10.192.230.192/26 |        64 | 10 (16%)   | 54 (84%)   |
| Block    | 10.192.230.64/26  |        64 | 24 (38%)   | 40 (62%)   |
| IP Pool  | 10.192.228.0/24   |       256 | 0 (0%)     | 256 (100%) |
| IP Pool  | 10.192.230.128/26 |        64 | 0 (0%)     | 64 (100%)  |
| IP Pool  | 10.192.229.0/24   |       256 | 0 (0%)     | 256 (100%) |
| IP Pool  | 10.192.231.0/24   |       256 | 0 (0%)     | 256 (100%) |
+----------+-------------------+-----------+------------+------------+
```

## Possible Solution

It's me for `nodeSelector: '!all()'` configuration wrong? I can't reserve the entire `pool-10` for manual allocation. `!all()` does not match any node, so pool-10's cidr will not be used for any automatic allocation.

## Steps to Reproduce (for bugs)

**1. pod cidr is `10.192.228.0/22`.**
**2. set ippool:** 
```
[root@master-23to24-199-146-28 ~]# kubectl get ippools.crd.projectcalico.org -oyaml
apiVersion: v1
items:
- apiVersion: crd.projectcalico.org/v1
  kind: IPPool
  metadata:
    name: default-pool
  spec:
    allowedUses:
    - Workload
    - Tunnel
    blockSize: 26
    cidr: 10.192.228.0/22
    ipipMode: Never
    nodeSelector: all()
    vxlanMode: Never
- apiVersion: crd.projectcalico.org/v1
  kind: IPPool
  metadata:
    name: pool-1
  spec:
    blockSize: 26
    cidr: 10.192.228.0/24
    ipipMode: Never
    natOutgoing: false
    nodeSelector: '!all()'
- apiVersion: crd.projectcalico.org/v1
  kind: IPPool
  metadata:
    name: pool-10
  spec:
    blockSize: 26
    cidr: 10.192.230.128/26
    ipipMode: Never
    natOutgoing: false
    nodeSelector: '!all()'
- apiVersion: crd.projectcalico.org/v1
  kind: IPPool
  metadata:
    name: pool-2
  spec:
    blockSize: 26
    cidr: 10.192.229.0/24
    ipipMode: Never
    natOutgoing: false
    nodeSelector: '!all()'
- apiVersion: crd.projectcalico.org/v1
  kind: IPPool
  metadata:
    name: pool-3
  spec:
    blockSize: 26
    cidr: 10.192.231.0/24
    ipipMode: Never
    natOutgoing: false
    nodeSelector: '!all()'
kind: List
metadata:
  resourceVersion: ""
```
**3. set IPReservation:** 
```
[root@master-23to24-199-146-28 ~]# kubectl get ipreservations.crd.projectcalico.org  -oyaml
apiVersion: v1
items:
- apiVersion: crd.projectcalico.org/v1
  kind: IPReservation
  metadata:
    name: my-ipreservation-1
  spec:
    reservedCIDRs:
    - 10.192.228.0/24
    - 10.192.229.0/24
    - 10.192.231.0/24
    - 10.192.230.0/25
    - 10.192.230.128/26
    - 10.192.230.192/27
kind: List
metadata:
  resourceVersion: ""
```
**4. Reserve ipamblocks.crd.projectcalico.org: `10-192-230-192-26`, delete others.**
```
[root@master-23to24-199-146-28 ~]# kubectl get ipamblocks.crd.projectcalico.org 
NAME                AGE
10-192-230-192-26   3h12m
[root@master-23to24-199-146-28 ~]# kubectl get blockaffinities.crd.projectcalico.org 
NAME                                                          AGE
hz-bd-k8s-test-bgp-node-23to24-199-146-50-10-192-230-192-26   3h14m
```

**5.  deploy 80 pods fill `10.192.230.224/27`:**
```
[root@master-23to24-199-146-28 ~]# kubectl get deployments.apps demo -oyaml
apiVersion: apps/v1
kind: Deployment
metadata:
  labels:
    app: demo
  name: demo
  namespace: default
spec:
  progressDeadlineSeconds: 600
  replicas: 70
  revisionHistoryLimit: 10
  selector:
    matchLabels:
      app: demo
  strategy:
    rollingUpdate:
      maxSurge: 25%
      maxUnavailable: 25%
    type: RollingUpdate
  template:
    metadata:
      creationTimestamp: null
      labels:
        app: demo
    spec:
      containers:
      - image: nginx:latest
        imagePullPolicy: IfNotPresent
        name: container-5l05o6
        ports:
        - containerPort: 80
          name: http-0
          protocol: TCP
        resources: {}
        terminationMessagePath: /dev/termination-log
        terminationMessagePolicy: File
      dnsPolicy: ClusterFirst
      restartPolicy: Always
      schedulerName: default-scheduler
      securityContext: {}
      serviceAccount: default
      serviceAccountName: default
      terminationGracePeriodSeconds: 30
status:
  availableReplicas: 30
  conditions:
  - lastTransitionTime: "2024-05-15T03:30:51Z"
    lastUpdateTime: "2024-05-15T03:30:51Z"
    message: Deployment has minimum availability.
    reason: MinimumReplicasAvailable
    status: "True"
    type: Available
  - lastTransitionTime: "2024-05-15T03:30:47Z"
    lastUpdateTime: "2024-05-15T03:30:52Z"
    message: ReplicaSet "demo-5956c7c885" has successfully progressed.
    reason: NewReplicaSetAvailable
    status: "True"
    type: Progressing
  observedGeneration: 1
  readyReplicas: 30
  replicas: 30
  updatedReplicas: 30
```
**6. Current block:** 
```
[root@master-23to24-199-146-28 ~]# kubectl get ipamblocks.crd.projectcalico.org 
NAME                AGE
10-192-230-192-26   3h22m
```

**7. chage IPReservation, delete `- 10.192.230.0/25`, `- 10.192.230.128/26`:**
```
[root@master-23to24-199-146-28 ~]# kubectl get ipreservations.crd.projectcalico.org my-ipreservation-1 -oyaml
apiVersion: crd.projectcalico.org/v1
kind: IPReservation
metadata:
  name: my-ipreservation-1
spec:
  reservedCIDRs:
  - 10.192.228.0/24
  - 10.192.229.0/24
  - 10.192.231.0/24
  - 10.192.230.192/27
```

**8. Now, get ipamblocks, found that more than `10-192-230-128-26` and `10-192-230-64-26`, at this point, there is no pod set `cni.projectcalico.org/ipv4pools: '["pool-10"]'` (PS: pool-10's cidr is `10.192.230.128/26`):**
```
[root@master-23to24-199-146-28 ~]# kubectl get ipamblocks.crd.projectcalico.org 
NAME                AGE
10-192-230-128-26   120m
10-192-230-192-26   3h22m
10-192-230-64-26    120m
```
**9. there is none pod with `cni.projectcalico.org/ipAddrs` or `cni.projectcalico.org/ipv4pools`**
```
[root@master-23to24-199-146-28 ~]# kubectl get pod -oyaml -A | egrep 'cni.projectcalico.org/ipAddrs|cni.projectcalico.org/ipv4pools'
[root@master-23to24-199-146-28 ~]# 
```

**10. You can see,  `10.192.230.128/26` already preempted.**

## Context

**After setting `IPPool.spec.nodeSelector: '!all()'`, must I manually specify `cni.projectcalico.org/ipv4pools: '[" poolName "]`, otherwise the calico-ipam cannot use this `poolName` to automatically assign `IP`**

## Your Environment

* Calico version
```
[root@master-23to24-199-146-28 ~]# calicoctl version
Client Version:    v3.22.3
Git commit:        1be7791d1
Cluster Version:   v3.22.3
Cluster Type:      kubespray,bgp,kubeadm,kdd,typha,k8s
```

* Orchestrator version (e.g. kubernetes, mesos, rkt):
```
[root@master-23to24-199-146-28 ~]# kubectl version
WARNING: This version information is deprecated and will be replaced with the output from kubectl version --short.  Use --output=yaml|json to get the full version.
Client Version: version.Info{Major:"1", Minor:"24", GitVersion:"v1.24.6", GitCommit:"b39bf148cd654599a52e867485c02c4f9d28b312", GitTreeState:"clean", BuildDate:"2022-09-21T13:19:24Z", GoVersion:"go1.18.6", Compiler:"gc", Platform:"linux/amd64"}
Kustomize Version: v4.5.4
Server Version: version.Info{Major:"1", Minor:"24", GitVersion:"v1.24.6", GitCommit:"b39bf148cd654599a52e867485c02c4f9d28b312", GitTreeState:"clean", BuildDate:"2022-09-21T13:12:04Z", GoVersion:"go1.18.6", Compiler:"gc", Platform:"linux/amd64"}
```

* Operating System and version:
* Link to your project (optional):

**Comment by caseydavenport** [comment]
@yell93 thanks for raising. To start, I recommend reading this: https://github.com/projectcalico/calico/issues/6412

TL;DR: the crd.projectcalico.org/v1 API is for internal use and isn't validated or defaulted, so you shouldn't use it. This may be why you are encountering issues using IP pools (note that the documentation uses "projectcalico.org/v3" everywhere). 

Similarly, you shouldn't create or delete the IPAMBlocks / BlockAffinity / IPAMHandles directly in normal operation. These are managed by Calico IPAM and manual intervention can lead to an inconsistent state. Probably not the cause of your issue, but may cause other problems down the road.

I don't think this is related to your issue, but Calico v3.22 is getting quite old at this point. I'd highly recommend upgrading to a more modern version of Calico.

**Comment by yell93** [comment]
> 谢谢你的提出。首先，我建议阅读以下内容：#6412
> 
> TL;DR：crd.projectcalico.org/v1 API 供内部使用，未经验证或默认，因此不应使用它。这可能就是您在使用 IP 池时遇到问题的原因（请注意，文档在任何地方都使用“projectcalico.org/v3”）。
> 
> 同样，您不应在正常操作中直接创建或删除 IPAMBlocks/BlockAffinity/IPAMHandles。这些由 Calico IPAM 管理，手动干预可能会导致状态不一致。可能不是您问题的原因，但可能会导致其他问题。
> 
> 我不认为这与您的问题有关，但 Calico v3.22 在这一点上已经很老了。我强烈建议升级到更现代的 Calico 版本。

Thank you for the reply, I will try using the new API and see how it goes with the upgraded version.

**Comment by yell93** [comment]
> crd.projectcalico.org/v

What impact will it have on my existing business if I `Enable kubectl to manage Calico APIs`, given that I have only used v1 in my cluster at present?
```
$ kubectl api-versions  |grep project
crd.projectcalico.org/v1
```

**Comment by yell93** [comment]
> 谢谢你的提出。首先，我建议阅读以下内容：#6412
> 
> TL;DR：crd.projectcalico.org/v1 API 供内部使用，未经验证或默认，因此不应使用它。这可能就是您在使用 IP 池时遇到问题的原因（请注意，文档在任何地方都使用“projectcalico.org/v3”）。
> 
> 同样，您不应在正常操作中直接创建或删除 IPAMBlocks/BlockAffinity/IPAMHandles。这些由 Calico IPAM 管理，手动干预可能会导致状态不一致。可能不是您问题的原因，但可能会导致其他问题。
> 
> 我不认为这与您的问题有关，但 Calico v3.22 在这一点上已经很老了。我强烈建议升级到更现代的 Calico 版本。

What impact will it have on my existing business if I , given that I have only used v1 in my cluster at present?Enable kubectl to manage Calico APIs
```
$ kubectl api-versions  |grep project
crd.projectcalico.org/v1
$ 
```

**Comment by caseydavenport** [comment]
If all of your resources are valid, it should have no impact. If your resources are invalid, it will help us to determine that.

You can also use `calicoctl` if you do not want to install the Calico API server in your cluster.

**Comment by sridhartigera** [comment]
@yell93 Any updates?
